File: 011_H-Index.cpp
/*
H-Index

Given an array of citation counts for a researcher, compute the researcher's h-index.
The h-index is the maximum h such that the researcher has at least h papers with >= h citations.

Strategy: Sort citations descending and find largest h where citations[h-1] >= h. O(n log n).
*/

#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

int hIndex(vector<int>& citations) {
    sort(citations.begin(), citations.end(), greater<int>());
    int h = 0;
    for (int i = 0; i < (int)citations.size(); ++i) {
        if (citations[i] >= i + 1) h = i + 1;
        else break;
    }
    return h;
}

int main() {
    vector<int> c1 = {3,0,6,1,5};
    cout << "H-Index 1: " << hIndex(c1) << '\n'; // 3

    vector<int> c2 = {10,8,5,4,3};
    cout << "H-Index 2: " << hIndex(c2) << '\n'; // 4

    vector<int> c3 = {0,0,0};
    cout << "H-Index 3: " << hIndex(c3) << '\n'; // 0

    return 0;
}

/*
Time Complexity: O(n log n)
Space Complexity: O(1) (ignoring input sort space)
*/

File: 011_H-Index.py
"""
H-Index

Compute the h-index given a list of citation counts.

Strategy: Sort descending and find largest h where citations[h-1] >= h.
"""

from typing import List

def h_index(citations: List[int]) -> int:
    citations.sort(reverse=True)
    h = 0
    for i, c in enumerate(citations):
        if c >= i + 1:
            h = i + 1
        else:
            break
    return h


if __name__ == '__main__':
    print('H-Index 1:', h_index([3,0,6,1,5]))
    print('H-Index 2:', h_index([10,8,5,4,3]))
    print('H-Index 3:', h_index([0,0,0]))

# Time Complexity: O(n log n)
# Space Complexity: O(1) (in-place sort)
